File: max_in_level.cpp
// TC: O(N) where N is the number of nodes in the tree;
// SC: O(depth of the tree)

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> largestValues(TreeNode* root) {
        queue<TreeNode*>q;
        vector<int>result;
        
        if(root == NULL)return result;
        q.push(root);
        while(!q.empty()){
            int lvl_size = q.size();
            int lvl_max = INT_MIN;
            for(int i = 0; i < lvl_size;i++){
                root = q.front();q.pop();
                lvl_max = max(lvl_max,root->val);
                if(root->left)q.push(root->left);
                if(root->right)q.push(root->right);
            }
            result.push_back(lvl_max);
        }
        return result;
        
    }
};

File: MaxInLevel.java
/*
For this solution, we utilise BFS by using a level order traversal to traverse the tree. At each level, we
maintain a variable that keeps a track of the maximum value at current level, as well as a variable for 
tracking the size of the current level. We add the maximum value found at each level to a list and return it.

Did this code run on leetcode: Yes
*/
class Solution {
    //Time Complexity: O(n)
    //Space Complexity: O(n)
    public List<Integer> largestValues(TreeNode root) {
        
        List<Integer> result = new ArrayList<>();
        
        if(root == null)
            return result;
        
        Queue<TreeNode> q = new LinkedList<>();
        
        q.add(root);
        
        while(!q.isEmpty())
        {
            int size = q.size();
            int max = Integer.MIN_VALUE;
            
            for(int i = 0; i < size; i++)
            {
                TreeNode curr = q.poll();
                max = Math.max(curr.val, max);
                
                if(curr.left != null)
                    q.add(curr.left);
                    
                if(curr.right != null)
                    q.add(curr.right);
            }
            
            result.add(max);
        }
        
        return result;
    }
}